kdc/kerberos5.c

```diff
@@ -695,68 +650,60 @@ pa_enc_chal_validate(astgs_request_t r, const PA_DATA *pa)
 	    krb5_free_error_message(r->context, msg);
 	    free(str);
 
+	    krb5_crypto_destroy(r->context, longtermcrypto);
 	    continue;
 	}
-	
-	ret = decode_PA_ENC_TS_ENC(ts_data.data,
-				   ts_data.length,
-				   &p,
-				   &size);
-	krb5_data_free(&ts_data);
-	if(ret){
-	    krb5_crypto_destroy(r->context, challangecrypto);
-	    ret = KRB5KDC_ERR_PREAUTH_FAILED;
-	    _kdc_r_log(r, 4, "Failed to decode PA-ENC-TS_ENC -- %s",
-		       r->cname);
-	    continue;
-	}
+    
+	/*
+	 * Found a key that the client used, lets pick that as the reply key
+	 */
 
-	if (labs(kdc_time - p.patimestamp) > r->context->max_skew) {
-	    char client_time[100];
+	krb5_free_keyblock_contents(r->context, &r->reply_key);
+	ret = krb5_copy_keyblock_contents(r->context, &k->key, &r->reply_key);
+	if (ret) {
+	    krb5_crypto_destroy(r->context, longtermcrypto);
+	    goto out;
+	}
 
-	    krb5_crypto_destroy(r->context, challangecrypto);
+	krb5_free_keyblock_contents(r->context, &challengekey);
 
-	    krb5_format_time(r->context, p.patimestamp,
-			     client_time, sizeof(client_time), TRUE);
+	/*
+	 * Provide KDC authentication to the client, uses a different
+	 * challenge key (different pepper).
+	 */
 
-	    ret = KRB5KRB_AP_ERR_SKEW;
-	    _kdc_r_log(r, 4, "Too large time skew, "
-		       "client time %s is out by %u > %u seconds -- %s",
-		       client_time,
-		       (unsigned)labs(kdc_time - p.patimestamp),
-		       r->context->max_skew,
-		       r->cname);
+	pepper1.data = "kdcchallengearmor";
+	pepper1.length = strlen(pepper1.data);
 
-	    free_PA_ENC_TS_ENC(&p);
+	ret = krb5_crypto_fx_cf2(r->context, r->armor_crypto, longtermcrypto,
+				 &pepper1, &pepper2, aenctype,
+				 &challengekey);
+	krb5_crypto_destroy(r->context, longtermcrypto);
+	if (ret)
 	    goto out;
-	}
 
-	free_PA_ENC_TS_ENC(&p);
+	ret = krb5_crypto_init(r->context, &challengekey, 0, &challengecrypto);
+	krb5_free_keyblock_contents(r->context, &challengekey);
+	if (ret)
+	    goto out;
 
-	ret = make_pa_enc_challange(r, challangecrypto);
-	krb5_crypto_destroy(r->context, challangecrypto);
+	ret = _krb5_make_pa_enc_challenge(r->context, challengecrypto,
+					  KRB5_KU_ENC_CHALLENGE_KDC,
+					  &r->outpadata);
+	krb5_crypto_destroy(r->context, challengecrypto);
 	if (ret)
 	    goto out;
 					    
 	set_salt_padata(&r->outpadata, k->salt);
-	krb5_free_keyblock_contents(r->context,  &r->reply_key);
-	ret = krb5_copy_keyblock_contents(r->context, &k->key, &r->reply_key);
-	if (ret)
-	    goto out;
 
-	/*
-	 * Success
-	 */
-	if (r->clientdb->hdb_auth_status)
-	    r->clientdb->hdb_auth_status(r->context, r->clientdb, r->client,
-					 HDB_AUTH_SUCCESS);
```

Comment:
Why is this lost?

Reply:
Fixed, although isn't that also called from `_kdc_as_rep()`? And, should errors be reported from the PKINIT and GSS PA code paths as well? Can this be subsumed into the auditing functions?

Reply:
Yes, see https://gitlab.com/samba-team/devel/lorikeet-heimdal/-/commit/b306eecbc9126b2afa5f83dd197296384a6506b9

It may well be that we remove this again in landing that.

Reply:
Yeah, I agree it looks like a duplicate now, sorry for not digging into it further.  Anyway, once this is in I'll clean up the patch to do this properly in auditing and ensure we do these things exactly once.